apps/examples/smartfs_powercut/smartfs_powercut_main.c

```diff
@@ -111,94 +122,162 @@ static int write_file(int fd, size_t bufsize)
+		len = read(fd, (void *)&g_buf, bufsize);
+		if (len != bufsize) {
+			/* If write is not finished properly */
+			printf("verify_file datachunk is not written fully..  path : %s ret : %d index : %d\n", path, len, i);
 			break;
 		} else {
+			/* Maybe this won't be happened but check crc for double check */
 			crc = crc16((const uint8_t *)g_buf.data, data_len);
 			if (crc != g_buf.crc) {
-				printf("verify_file %s crc mismatch!! calulcated crc : %d written crc : %d read count : %d\n", path, crc, g_buf.crc, i);
+				printf("verify_file crc mismatch!! path : %s calulcated crc : %d written crc : %d index : %d\n", path, crc, g_buf.crc, i);
 				break;
 			}
 		}
 	}
+
+	*index = i;
+
+	/* shrink to last valid index here, it is necessary to bury invalid data chunk */
+	if ((i > 0) && (i < TEST_WRITECOUNT)) {
+		/* truncate current file to include valid data only */
+		size = i * bufsize;
+		
+		ret = lseek(fd, 0, SEEK_SET);
+		if (ret != OK) {
+			printf("verify_file, seek to 0 failed\n");
+			ret = ERROR;
+			goto error_with_fd;
+		}
+
+		ret = ftruncate(fd, size);
+		if (ret != OK) {
+			printf("Shrink failed filename : %s target size : %d errno : %d\n", path, size, errno);
+			ret = ERROR;
+			goto error_with_fd;
+		}
+
+		ret = stat(path, &st);
+		if (ret == OK) {
+			if (st.st_size != size) {
+				printf("Shrunk size is wrong. filename : %s target size : %d file size : %d\n", path, size, st.st_size);
+				ret = ERROR;
+				goto error_with_fd;
+			}
+			printf("shrink works properly, target size : %d current size : %d\n", size, st.st_size);
+		} else {
+			printf("stat error. path : %d errno : %d\n", path, errno);
+			ret = ERROR;
+			goto error_with_fd;
+		}
+	}
+
+	ret = OK;
+
+error_with_fd:
 	close(fd);
+	return ret;
+}
 
-	printf("%s last valid write count : %d\n", path, i);
-	return i;
+static int verify_file(char *path, int bufsize, int *idx, file_type_t type)
+{
+	if (type == FILE_TYPE_RANDOM) {
+		return verify_random_file(path, bufsize, idx);
+	}
+	/* TODO verify proper text type */
+	return OK;
 }
 
-int recovery_file(char *src, char* backup, char* backup_dir, int bufsize, int idx)
+static int recovery_random_file(char *src, char* backup, char* backup_dir, int bufsize, int idx)
 {
 	int fd1;
 	int fd2;
 	int ret;
+	struct stat st;
 	off_t offset;
 	off_t start_offset;
+	int filesize;
 	
-	/* If idx is ERROR, it means it didn't create src file so bring backup file */
 	printf("recovery_file src : %s idx : %d\n", src, idx);
-	if (idx == ERROR || idx == 0 || idx == TEST_WRITECOUNT) {
+	/* If index is FILE_CHUNK_NO_INDEX, There is no tested file, so just rollback backup file
+	 * If index is 0, we should recovery whole of data from back up file, so just rollback it.
+	 * If idnex is same as _TEST_WRITECOUNT, totally written, so no recovery necessary.
+	 */
+	if (idx == FILE_CHUNK_NO_INDEX || idx == 0 || idx == TEST_WRITECOUNT) {
```

Comment:
in case `idx == TEST_WRITECOUNT`
We unlink existing file which is complete?
Is that necessary??

Why do we call recovery if index = TEST_WRITECOUNT?

Reply:
Please make sure that is application's role, it is testing. As possible as it can, it should test many thing.
if idx is same as TEST_WRITECOUNT, it means that power has cut off after write is finished, and CRCs are fine.
In this case, we don't have to recovery.
But make sure that we have to TEST anything , in this case It RENAME backup to src, and filanny src to backup, so we can test 2 time of rename.